From bdf8f49962174237b86cda18f59f294f43756005 Mon Sep 17 00:00:00 2001
From: Amith A <quic_amitajit@quicinc.com>
Date: Tue, 20 Aug 2024 14:35:07 +0530
Subject: [PATCH] UPSTREAM: hostapd: Extend frequency configuration to handle 6
 GHz channel 2

In hostapd_set_freq_params(), if center_segment0 is 2, call
ieee80211_chan_to_freq() with operating class 136 instead of 131.

This is needed because, channel 2 is an exception in the 6 GHz band. It
comes before channel 1 and is part of operating class 136.

Channels order in 6 GHz:
    2 (Operating Class 136)
    1   5   9 ....  (Operating Class 131)

Signed-off-by: Amith A <quic_amitajit@quicinc.com>
---
 src/common/hw_features_common.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/common/hw_features_common.c b/src/common/hw_features_common.c
index 6df1897..68d03fc 100644
--- a/src/common/hw_features_common.c
+++ b/src/common/hw_features_common.c
@@ -452,6 +452,7 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 		} else {
 			int freq1, freq2 = 0;
 			int bw = center_idx_to_bw_6ghz(center_segment0);
+			int opclass;
 
 			if (bw < 0) {
 				wpa_printf(MSG_ERROR,
@@ -459,7 +460,10 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 				return -1;
 			}
 
-			freq1 = ieee80211_chan_to_freq(NULL, 131,
+			/* The 6 GHz channel 2 uses a different operating class
+			 */
+			opclass = center_segment0 == 2 ? 136 : 131;
+			freq1 = ieee80211_chan_to_freq(NULL, opclass,
 						       center_segment0);
 			if (freq1 < 0) {
 				wpa_printf(MSG_ERROR,
@@ -475,7 +479,7 @@ int hostapd_set_freq_params(struct hostapd_freq_params *data,
 					return -1;
 				}
 
-				freq2 = ieee80211_chan_to_freq(NULL, 131,
+				freq2 = ieee80211_chan_to_freq(NULL, opclass,
 							       center_segment1);
 				if (freq2 < 0) {
 					wpa_printf(MSG_ERROR,
-- 
2.34.1

